src/dawn/AST/ASTStmt.h

```diff
@@ -141,11 +185,23 @@ class BlockStmt : public Stmt {
 
   template <class Iterator>
   void insert_back(Iterator begin, Iterator end) {
+    std::for_each(begin, end, [&](const std::shared_ptr<Stmt>& stmt) {
+      DAWN_ASSERT(stmt);
+      DAWN_ASSERT_MSG((checkSameDataType(*stmt)),
+                      "Trying to insert child Stmt with different data type");
+    });
     statements_.insert(statements_.end(), begin, end);
   }
 
-  void push_back(const std::shared_ptr<Stmt>& stmt) { statements_.push_back(stmt); }
+  void push_back(const std::shared_ptr<Stmt>& stmt) {
+    DAWN_ASSERT(stmt);
+    DAWN_ASSERT_MSG((checkSameDataType(*stmt)),
+                    "Trying to insert child Stmt with different data type");
+    statements_.push_back(stmt);
+  }
 
+  // TODO refactor_AST: this non-const getter is a source of problems: no runtime checks on data
+  // type when user changes the vector!
```

Comment:
Unrelated, but all the setters in Stmt and Expr are a problem anyway (if not needed) because it gives the impression that everything can be changed.

Reply:
The state of Stmts and Exprs classes right now is that they have very little logic inside. In the future we may want to restrict the way they can be changed by exposing an API that allows only certain operations.